components/applications-service/pkg/storage/postgres/service_groups.go

```diff
@@ -21,6 +21,56 @@ import (
 // OK        | Else if all services are 'Ok'
 // ---------------------------------------------------------------------
 const (
+	// @afiune IDEA: Bring back the main query. The explanation why is below.
```

Comment:
So, I realise it's not beautiful to duplicate, **but**: If we get rid of the nesting in "the big query", can't we **use the view, and _restrict `SELECT`s of that view_ in the same way**?

I think this might be largely equivalent (except for column order):
```sql
CREATE OR REPLACE VIEW service_group_health AS
  SELECT  sg.id AS id
         ,sg.deployment_id AS deployment_id
         ,sg.name AS name
         ,COUNT(*) FILTER (WHERE s.health = 'OK') AS health_ok
         ,COUNT(*) FILTER (WHERE s.health = 'CRITICAL') AS health_critical
         ,COUNT(*) FILTER (WHERE s.health = 'WARNING') AS health_warning
         ,COUNT(*) FILTER (WHERE s.health = 'UNKNOWN') AS health_unknown
         ,COUNT(*) AS health_total
         ,round(COUNT(*) FILTER (WHERE s.health='OK') * 100 / COUNT(*)) AS percent_ok
         ,(CASE WHEN (COUNT(*) FILTER (WHERE s.health='OK')) > 0 THEN '1_CRITICAL'
                WHEN (COUNT(*) FILTER (WHERE s.health='UNKNOWN')) > 0 THEN '2_UNKNOWN'
                WHEN (COUNT(*) FILTER (WHERE s.health='WARNING')) > 0 THEN '3_WARNING'
                ELSE '4_OK' END ) AS health
         ,(SELECT array_agg(DISTINCT CONCAT (s.package_ident))
             FROM service AS s
            WHERE s.group_id = sg.id) AS releases
         ,d.app_name AS app_name
         ,d.environment AS environment
        FROM service_group AS sg
        JOIN service AS s
             ON s.group_id = sg.id
        JOIN deployment AS d
             ON sg.deployment_id = d.id
       GROUP BY sg.id, sg.deployment_id, sg.name, d.app_name, d.environment;
```

I believe you should then be able to

```sql
SELECT * FROM service_group_health sgh WHERE sgh.id IN (SELECT DISTINCT group_id FROM service WHERE origin = 'custom' AND name = 'nginx')
```

etc. _I might be missing something_.

Notes:
- `COUNT(s.health) FILTER (WHERE s.health='yadda')` is the same as `COUNT(*) FILTER (WHERE s.health='yadda')` -- `COUNT(s.health)` would be different from `COUNT(*)` if `s.health` coud be NULL, but it cannot with that filter
- for the percentage, is better to multiply _first_ and divide _second_ ([see here](https://dba.stackexchange.com/a/69132))

Reply:
You are right Stephan, this could be another approach but, now that you mention it, I think we can't use this approach because we still need to reduce services counts, that is, to display service-groups with information coming from the filteres services. I personally don't think is the right thing to do to hide or reduce the service-groups but I think this is what we want to have for iteration number one.

Example:
service group **redis.default** has four services:
1) 2 services with site equals to `us`
1) 2 services with site equals to `eu`

If I do a filter `site=eu`, then these things should happen:
1) display only service groups that has services with the site tag equal to `eu`
1) reduce the service group health summary to only count such services. (in this case the service group should report only two services instead of four.)

The approach that @danielsdeleo had was to use [`WITH` statements](https://www.postgresql.org/docs/9.1/queries-with.html), which now I think might be the best approach, though, we still need to move the view back to the code.

Reply:
Do we need to use _one view_ (or function or whatever) to support those different kinds of queries? I suppose there could be a view that has less (or no) aggregation happening, and another one that does aggregate services; queriers could use one or the other, depending on what they need to know/filter.